core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/entity/descriptor/EntityDescriptor.java

```diff
@@ -684,22 +687,57 @@ public long getMaximumValueCount(Solution_ solution, Object entity) {
 
     }
 
-    public long getProblemScale(Solution_ solution, Object entity) {
-        int genuineEntityCount = getSolutionDescriptor().getGenuineEntityCount(solution);
-        long problemScale = 1L;
+    public void processProblemScale(ScoreDirector<Solution_> scoreDirector, Solution_ solution, Object entity,
+            SolutionDescriptor.ProblemScaleTracker tracker) {
         for (GenuineVariableDescriptor<Solution_> variableDescriptor : effectiveGenuineVariableDescriptorList) {
             long valueCount = variableDescriptor.getValueRangeSize(solution, entity);
-            problemScale *= valueCount;
-            if (variableDescriptor.isListVariable()) {
-                // This formula probably makes no sense other than that it results in the same problem scale for both
-                // chained and list variable models.
-                // TODO fix https://issues.redhat.com/browse/PLANNER-2623 to get rid of this.
-                problemScale *= valueCount;
-                problemScale /= genuineEntityCount;
-                problemScale += valueCount;
+            // TODO: When minimum Java supported is 21, this can be replaced with a sealed interface switch
+            if (variableDescriptor instanceof BasicVariableDescriptor<Solution_> basicVariableDescriptor) {
+                if (basicVariableDescriptor.isChained()) {
+                    // An entity is a value
+                    tracker.listTotalValueCount().increment();
+                    if (!isMovable(scoreDirector, entity)) {
+                        tracker.listPinnedValueCount().increment();
+                    }
+                    // Anchors are entities
+                    ValueRange<?> valueRange = variableDescriptor.getValueRangeDescriptor().extractValueRange(solution, entity);
+                    if (valueRange instanceof CountableValueRange<?> countableValueRange) {
+                        Iterator<?> valueIterator = countableValueRange.createOriginalIterator();
+                        while (valueIterator.hasNext()) {
+                            Object value = valueIterator.next();
+                            if (variableDescriptor.isValuePotentialAnchor(value)) {
+                                if (tracker.visitedAnchorSet().contains(value)) {
+                                    continue;
+                                }
+                                tracker.visitedAnchorSet().add(value);
+                                tracker.listTotalEntityCount().increment();
+                                tracker.listMovableEntityCount().increment();
+                                // Assumes anchors are not pinned
+                            }
+                        }
+                    } else {
+                        // Assume 0 entities (this might be impossible)
```

Comment:
Uncountable value ranges already fail fast elsewhere. Please confirm and if that's the case, please also fail fast here. People shouldn't be using them.